### PR TITLE
add unused_top_members_in_executable_libraries

### LINKED ISSUE #2395: Report unused public elements in an entrypoint library
**Describe the rule you'd like to see implemented**

If a library has an entrypoint, I'd like to report any publicly accessible elements which are not used within the library as unused.

An entrypoint is a top-level `main` function, or a function annotated with `@pragma(vm:entry-point)`. These files are typically:

* a VM script in `bin/`,
* a web app in `web/`,
* a VM tool in `tool/`,
* a test like `test/**/*_test.dart`,

I think there will be very very few false positives. And perhaps any false positive indicates a code smell in library arrangement (the analyzer codebase I believe has some test files which reference another test file, each of which has a `main`).

Any element which is _not_ publicly visible should already be reported with analyzer's `unused_local_variable`, `unused_element`, or `unused_field` diagnostics. This lint would cover publicly visible elements:

* Any public class instance members (fields, methods, constructors), even those on private classes and mixins.
* Any public top-level functions, variables, classes, enums, mixins, extensions, and typedefs.

**Examples**

```dart
// foo_test.dart

void main() {
  test(...);
}

// Unused class
class MockStream extends Mock implements Stream {
  // Unused field
  int y = 7;

  // Unused method
  void m() {}
}

// Unused top-level function
void helper() {}

// Unused top-level variable
const x = 1;
```

**Additional context**

This is an alternative route to #685 and #1258.

### PR DESCRIPTION
**Pull request by a14n**
# Description

Report unused top-level members in executable libraries.

Fixes #2395

### CODE REVIEW

**Review by bwilkerson — COMMENTED** [review]
It isn't clear to me that this is a rule we want to have.

As noted below, it will have false positives unless a convention is followed, and there's no support for enforcing that convention. If we think that this is a convention that we should support, then we should probably have a lint that ensures that it isn't violate. It seems to me that without that enforcement we run a risk of having a high rate of false positives from this rule.

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart:18]: I added a comment regarding the assumption.

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart:35]: done

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart:121]: good catch. Done

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart:123]: Correct. Done.

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart:165]: Actually I changed the test to check if there is a declaration conforming the `_isEntryPoint` function (inlined).

**Review by a14n — COMMENTED** [review]
I ran this lint on flutter (see https://github.com/flutter/flutter/pull/107464) and I didn't face false positives.

**Comment by a14n** [comment]
PTAL

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (srawlins) [inline test_data/rules/unused_top_members_in_executable_libraries.dart:5]: Because I'm not familiar with `traverseNodesInDFS`, and therefore don't trust it implicitly :), I'd love to see some interesting cases like:

```dart
class C {
  const C();
}

void f<T extends C> {} // Usage in type bound

void Function(C c) f2() {} // Usage in Function type

void f3([Object p = C()]) {} // Usage in default value

@C() // Usage in annotation
void f4() {}

extension on Future<C> {} // Usage in type parameter in extension `on` clause.

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (srawlins) [inline test_data/rules/unused_top_members_in_executable_libraries.dart:6]: Here's a case where we _do_ report an element as unused, even though it is _referenced_: anywhere in the right operand of `is`, _except_ for type aliases:

```dart
class C {}

typedef ListInt = List<int>;

void f(Object p) {
  if (p is C || p is Future<C>) {} // NOT usage
  if (p is ListInt) {} // YES usage
}
```

Comment (srawlins) [inline test_data/rules/unused_top_members_in_executable_libraries.dart:6]: If a variable declaration uses a type but not in an initializer, that is _not_ use:

```dart
class C {}

C? c; // NOT usage
```

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (srawlins) [inline test_data/rules/unused_top_members_in_executable_libraries.dart:26]: Could you add two constructors here to show that the presence of constructors (which refer to the name of the class) does not count the class as "used"

```
class C { // LINT
  C();
  C.named();
}
```

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (srawlins) [inline test_data/rules/unused_top_members_in_executable_libraries.dart:50]: Could you add tests showing that assignment is not "usage." Something like:

```dart
int? id2 = 0;
void f() {
  id2 = 1; // NOT usage
  id2 ??= 2; // NOT usage
}
```

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (srawlins) [inline test_data/rules/unused_top_members_in_executable_libraries.dart:6]: How do you want to treat doc comments? I think they are probably _not_ use:

```
class C {}

/// I love [C].
void main() {}
```

Comment (srawlins) [inline test_data/rules/unused_top_members_in_executable_libraries.dart:6]: Could you add tests for both a named an an unnamed extension?

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (srawlins) [inline test_data/rules/unused_top_members_in_executable_libraries.dart:50]: Could you add tests similar to this `id` field one but with setters and getters?

```dart
int get id3 => 0;
void set id3(int value) {};

void f() {
  id3 += 4; // usage
}

**Review by srawlins — CHANGES_REQUESTED** [review]

**Comment by srawlins** [comment]
@pq is there a way to benchmark this before it is landed? I worry about performance on this one because of traverseNodesInDFS.

**Comment by pq** [comment]
> is there a way to benchmark this before it is landed?

Great question!  The good news is each PR triggers the benchmark bot and that gives us a feel.  If you expand the checks, you'll see it:

image


Click on details and you can look at the output.

In this case, you're right that as implemented, this is relatively costly:

image

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart]: done

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart]: done

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart:71]: done

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart:103]: Without type annotations it becomes a `Map<dynamic, Set<Declaration>>` instead of a `Map<Declaration, Set<Declaration>>`

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart]: done

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart]: I move this code on the first check.

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart]: done

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart:156]: > we can perhaps delay this if we know that `elementAnnotation.element` is not the `pragma` class?

`elementAnnotation.element` would point to the ConstructorElement and it will work most of the time. Unfortunatelly I think this wouldn't support code like:

```dart
const entryPoint = pragma('vm:entry-point');

@entryPoint
f(){}
```

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply test_data/rules/unused_top_members_in_executable_libraries.dart]: done

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:5]: I added those tests

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:59]: done

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:5]: I may be wrong but some of the comments you did was not really related to this lint (warn on **public** top level members that are not accessible directly or transitivelly from an entry-point ) and so I didn't add test case for those comments I left unanswered.

Comment (srawlins) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:5]: 🎉 Thanks!

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (srawlins) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:6]: Er, I guess not unnamed, because that is handled by the `unused_element` hint.

Comment (srawlins) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:6]: Is this one of the cases you thought would be unrelated? I'm not sure why. In this case, I would want C reported as unused, and I would not want ListInt reported as unused.

Comment (a14n) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:6]: TBH I don't understand why `C` should be reported here (assuming `f` is reachable from an entrypoint).

The rule doesn't really analyze the execution flow, it only visit the AST children of the `f` node.

Comment (a14n) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:6]: Named extension is now covered by https://github.com/dart-lang/linter/pull/3513/files#diff-4f5160f7cce759c5604eb650859c439c3084aee4efd0b42cdd07d2bb8d3f69f1R109

Comment (a14n) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:6]: Perhaps the term _unused_ is a little misleading. It should be understood as _unaccessible from any entry point_.

Comment (srawlins) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:6]: I don't think execution flow is important either; just syntax tree.

In this example, C would be considered _used_, right?

```dart
class C {}

void f() {
  var c = C();
}
```

or even

```dart
class C {}

void f(C c) {}
```

I think those would be "usages" of C. However, in this code:

```dart
class C {}

void f(Object p) {
  if (p is C || p is Future<C>) {}
}
```

I would not count C as being "used." (This is how unused_element works.) Any time `C` is found in the right side of an `is ` expression does not count as usage.

Comment (a14n) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:6]: > I would not count C as being "used." (This is how unused_element works.) Any time `C` is found in the right side of an `is ` expression does not count as usage.

Thanks for the explanation. 

As mentioned in my in-between comment above my lint don't really mimics the `unused_element` behaviour. It simply visit recursivelly the whole AST (without exception like in the example you provided with right side of `is`) starting from entrypoints and warns if a top public declaration is not reachable.

Comment (srawlins) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:6]: Would `traverseNodesInDFS`'s `excludeCriteria` parameter allow this? I can't tell offhand because I can't find any tests for `traverseNodesInDFS`.

Comment (a14n) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:6]: I could add this kind of rules but I don't know where to find a list of all the exceptions to mimic the behaviour of `unused_elements`. 
That said I'm not sure the cost of maintaining this logic in 2 places is worth it. If some changes land on `unused_element` we will have to redo the job on linter side.
(Side note: I'm a little surprise that `C` is not considered used because if you remove `class C {}` the code doesn't compile)

Comment (bwilkerson) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:6]: > Perhaps the term _unused_ is a little misleading. It should be understood as _unaccessible from any entry point_.

If what you're really trying to lint are declarations that are not reachable from `main`, then yes, the terminology is getting in the way. Consider changing the name of the lint to `unreachable_from_main`.

And if that's the case, then we can re-review this PR from that perspective (because it changes the kinds of possible bugs that we should be looking for).

Comment (bwilkerson) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:6]: > I'm a little surprise that `C` is not considered used because if you remove `class C {}` the code doesn't compile

I have to agree, that does seem a bit odd.

From a tree-shaking perspective, you couldn't remove `C` unless you also removed the then-branch of the `if`. If we think that there can't be any instances of `C` or any subclass of `C`, then it seems like the `if` should be marked as dead code, but `C` shouldn't be reported until the `if` is removed. (But maybe I'm misunderstanding your example.)

Comment (a14n) [reply test_data/rules/unused_top_members_in_executable_libraries.dart:6]: Sorry for the bad naming :-/ I used this name because of #2395 that has this wording.

> Consider changing the name of the lint to `unreachable_from_main`.

:+1: I will update the PR

Comment (a14n) [reply test_data/rules/unused_top_members_in_executable_libraries.dart]: Renamed. PTAL

File: lib/src/rules/unused_top_members_in_executable_libraries.dart
Comment (srawlins) [reply lib/src/rules/unused_top_members_in_executable_libraries.dart]: Ack

File: lib/src/rules/unreachable_from_main.dart
Comment (bwilkerson) [inline lib/src/rules/unreachable_from_main.dart:146]: Given that these are declarations, it would be better if we could report the lint on the name of the declaration, or the `extension` keyword in the case of an unnamed extension.

Comment (a14n) [reply lib/src/rules/unreachable_from_main.dart]: done

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (srawlins) [reply test_data/rules/unused_top_members_in_executable_libraries.dart]: > I could add this kind of rules but I don't know where to find a list of all the exceptions to mimic the behaviour of unused_elements.

I've just enumerated the exceptions in the docs: https://github.com/dart-lang/sdk/commit/2bedf1e55d593ed88cb63bb9321354ccb6e0bc15

> If we think that there can't be any instances of C or any subclass of C, then it seems like the if should be marked as dead code, ...

Filed https://github.com/dart-lang/sdk/issues/49584

> ... but C shouldn't be reported until the if is removed.

That's fair. We report it today and I don't think we should stop reporting it unless we complete #49584 as it provides value.

File: lib/src/rules/unreachable_from_main.dart
Comment (srawlins) [inline lib/src/rules/unreachable_from_main.dart:19]: I think it would be good to define "executable library" in the description.

File: test_data/rules/unreachable_from_main.dart
Comment (srawlins) [inline test_data/rules/unreachable_from_main.dart:28]: Can you add a test for multiple variables declared in a single declaration? I guess something like at least one of them is reachable, and at least one is not.

File: lib/src/rules/unreachable_from_main.dart
Comment (srawlins) [inline lib/src/rules/unreachable_from_main.dart:116]: I think this `map` callback is equivalent to some invocation of `Element.thisOrAncestorMatching`, which would make it more readable. Something like

```dart
.map((e) =>
    e.thisOrAncestorMatching(
        (a) => a.enclosingElement2 == null || a.enclosingElement2 is CompilationUnitElement))
```

maybe? WDYT?

File: lib/src/rules/unreachable_from_main.dart
Comment (srawlins) [inline lib/src/rules/unreachable_from_main.dart:130]: I think this constructs a `Queue<dynamic>`. We want `Queue.of`.

Comment (srawlins) [inline lib/src/rules/unreachable_from_main.dart:130]: Could you add a comment above this declaration explaining what `toTraverse` is? Or perhaps what the algorithm in the while-loop is.

File: lib/src/rules/unreachable_from_main.dart
Comment (srawlins) [inline lib/src/rules/unreachable_from_main.dart:167]: 5 levels of if-statements makes this code pretty hard for me to read. Can we instead short circuit `return false` when we hit a condition that makes this _not_ a pragma, and flatten out the code?

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (srawlins) [reply test_data/rules/unused_top_members_in_executable_libraries.dart]: This comment still stands. I want to make sure that `traverseNodeInDFS` and your algorithm do not count `C.named();` as a legitimate reachability reference to `C`. `C` should be marked unreachable.

File: lib/src/rules/unreachable_from_main.dart
Comment (srawlins) [inline lib/src/rules/unreachable_from_main.dart:103]: I think it would be good to have a comment above this variable. It takes a lot of logic to figure out what these 25 lines do.

**Review by srawlins — CHANGES_REQUESTED** [review]

File: lib/src/rules/unreachable_from_main.dart
Comment (a14n) [reply lib/src/rules/unreachable_from_main.dart:19]: Description added.

File: test_data/rules/unreachable_from_main.dart
Comment (a14n) [reply test_data/rules/unreachable_from_main.dart:29]: done

File: lib/src/rules/unreachable_from_main.dart
Comment (a14n) [reply lib/src/rules/unreachable_from_main.dart]: done

File: lib/src/rules/unreachable_from_main.dart
Comment (a14n) [reply lib/src/rules/unreachable_from_main.dart]: Good call! Isn't there a lint for that :) ?

Comment (a14n) [reply lib/src/rules/unreachable_from_main.dart]: done

File: lib/src/rules/unreachable_from_main.dart
Comment (a14n) [reply lib/src/rules/unreachable_from_main.dart:107]: done

File: test_data/rules/unused_top_members_in_executable_libraries.dart
Comment (a14n) [reply test_data/rules/unused_top_members_in_executable_libraries.dart]: done

**Comment by a14n** [comment]
PTAL

File: lib/src/rules/unreachable_from_main.dart
Comment (srawlins) [reply lib/src/rules/unreachable_from_main.dart]: I had the same question, haha. https://github.com/dart-lang/linter/issues/2555

**Comment by a14n** [comment]
I perhaps missed some comments. Is there something blocking on this PR?

**Comment by srawlins** [comment]
I forgot to click Send on my last 3 comments; sorry about that! I think my open comments amount to:

* How do we treat references in comments?
* Can we avoid evaluating the `@pragma` constant until we know its a `@pragma` annotation?

**Comment by a14n** [comment]
> * How do we treat references in comments?

There was already a test to trigger a diagnostic on elements only ref in comment (see `Comment` in test file)

> * Can we avoid evaluating the `@pragma` constant until we know its a `@pragma` annotation?

Done. Thanks for the code snippet.

**Comment by a14n** [comment]
After rebasing on `main` (and upgrading the analyzer) references in comments are now visited. I updated the test to not mark `Comment` as unreachable.

PTAL

File: lib/src/rules/unreachable_from_main.dart
Comment (srawlins) [inline lib/src/rules/unreachable_from_main.dart:23]: I think there is a stray `)` here, at the end of the sentence.

**Review by srawlins — APPROVED** [review]
Looks great! Thanks! 🎉

**Pull request merged**
